// webpack-bundle-analyzer/lib/tree/ConcatenatedModule.js
import { __exports as ConcatenatedModule$1 } from '../../../../_virtual/ConcatenatedModule.js';
import '../../../lodash/lodash.js';
import './Module.js';
import './ContentModule.js';
import './ContentFolder.js';
import './utils.js';
import { __exports as Module } from '../../../../_virtual/Module.js';
import { __exports as ContentModule } from '../../../../_virtual/ContentModule.js';
import { __exports as ContentFolder } from '../../../../_virtual/ContentFolder.js';
import { __exports as utils } from '../../../../_virtual/utils2.js';
import { l as lodash } from '../../../../_virtual/lodash.js';

Object.defineProperty(ConcatenatedModule$1, "__esModule", {
  value: true
});
ConcatenatedModule$1.default = void 0;

var _lodash = _interopRequireDefault(lodash.exports);

var _Module = _interopRequireDefault(Module);

var _ContentModule = _interopRequireDefault(ContentModule);

var _ContentFolder = _interopRequireDefault(ContentFolder);

var _utils = utils;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ConcatenatedModule extends _Module.default {
  constructor(name, data, parent) {
    super(name, data, parent);
    this.name += ' (concatenated)';
    this.children = Object.create(null);
    this.fillContentModules();
  }

  fillContentModules() {
    this.data.modules.forEach(moduleData => this.addContentModule(moduleData));
  }

  addContentModule(moduleData) {
    const pathParts = (0, _utils.getModulePathParts)(moduleData);

    if (!pathParts) {
      return;
    }

    const [folders, fileName] = [pathParts.slice(0, -1), _lodash.default.last(pathParts)];
    let currentFolder = this;
    folders.forEach(folderName => {
      let childFolder = currentFolder.getChild(folderName);

      if (!childFolder) {
        childFolder = currentFolder.addChildFolder(new _ContentFolder.default(folderName, this));
      }

      currentFolder = childFolder;
    });
    const module = new _ContentModule.default(fileName, moduleData, this);
    currentFolder.addChildModule(module);
  }

  getChild(name) {
    return this.children[name];
  }

  addChildModule(module) {
    module.parent = this;
    this.children[module.name] = module;
  }

  addChildFolder(folder) {
    folder.parent = this;
    this.children[folder.name] = folder;
    return folder;
  }

  mergeNestedFolders() {
    _lodash.default.invokeMap(this.children, 'mergeNestedFolders');
  }

  toChartData() {
    return { ...super.toChartData(),
      concatenated: true,
      groups: _lodash.default.invokeMap(this.children, 'toChartData')
    };
  }

}

ConcatenatedModule$1.default = ConcatenatedModule;
