// webpack-bundle-analyzer/lib/tree/ContentModule.js
import { __exports as ContentModule$1 } from '../../../../_virtual/ContentModule.js';
import './Module.js';
import { __exports as Module } from '../../../../_virtual/Module.js';

Object.defineProperty(ContentModule$1, "__esModule", {
  value: true
});
ContentModule$1.default = void 0;

var _Module = _interopRequireDefault(Module);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ContentModule extends _Module.default {
  constructor(name, data, ownerModule, parent) {
    super(name, data, parent);
    this.ownerModule = ownerModule;
  }

  get parsedSize() {
    return this.getSize('parsedSize');
  }

  get gzipSize() {
    return this.getSize('gzipSize');
  }

  getSize(sizeType) {
    const ownerModuleSize = this.ownerModule[sizeType];

    if (ownerModuleSize !== undefined) {
      return Math.floor(this.size / this.ownerModule.size * ownerModuleSize);
    }
  }

  toChartData() {
    return { ...super.toChartData(),
      inaccurateSizes: true
    };
  }

}

ContentModule$1.default = ContentModule;
